cmd/libsnap-confine-private/mountinfo.c

```diff
@@ -126,42 +127,86 @@ static void show_buffers(const char *line, int offset,
 	fprintf(stderr, ">%s<\n", line);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i) {
+	for (size_t i = 0; i < strlen(line); ++i) {
 		int c = entry->line_buf[i];
 		fputc(c == 0 ? '@' : c == 1 ? '#' : c, stderr);
 	}
 	fputc('<', stderr);
 	fputc('\n', stderr);
 
 	fputc('>', stderr);
-	for (int i = 0; i < strlen(line); ++i)
+	for (size_t i = 0; i < strlen(line); ++i)
 		fputc('=', stderr);
 	fputc('<', stderr);
 	fputc('\n', stderr);
 #endif				// MOUNTINFO_DEBUG
 }
 
+static bool is_octal_digit(char c)
+{
+	return c >= '0' && c <= '7';
+}
+
 static char *parse_next_string_field(sc_mountinfo_entry * entry,
-				     const char *line, int *offset)
+				     const char *line, size_t * offset)
 {
-	int offset_delta = 0;
-	char *field = &entry->line_buf[0] + *offset;
-	if (line[*offset] == ' ') {
-		// Special case for empty fields which cannot be parsed with %s.
-		*field = '\0';
-		*offset += 1;
-	} else {
-		int nscanned =
-		    sscanf(line + *offset, "%s%n", field, &offset_delta);
-		if (nscanned != 1)
-			return NULL;
-		*offset += offset_delta;
-		if (line[*offset] == ' ') {
-			*offset += 1;
+	const char *input = &line[*offset];
+	char *output = &entry->line_buf[*offset];
+	const char *s;		// Used for escape logic.
+	size_t input_idx = 0;
+	size_t output_idx = 0;
+
+	// Scan characters until we run out of memory to scan or we find a space.
+	// The kernel uses simple octal escape sequences for the following:
+	// space, tab, newline, backwards slash. Everything else is copied
+	// verbatim.
+	bool stop = false;
+	for (int c; c = input[input_idx], c != '\0' && !stop;
+	     input_idx++, output_idx++) {
+		switch (c) {
+		case ' ':
+			// Fields are space delimited. Represent the space
+			// as the end-of-string marker, skip over it, and stop parsing.
+			output[output_idx] = '\0';
```

Comment:
It's not a problem but it looks like this is redundant with the assignment after the for loop, but this importantly needs to be done here otherwise there is an off-by-one. Ie:
```
36 35 98:0 /mnt1 /mnt2
```
When the space after /mnt1 is found, output_idx is 5, so you put a NUL at output[5] and flag to stop. The loop continues, increments output_idx to 6 then terminates the loop. At this point you add a second NUL at output[6].

This is all perfectly safe and not a problem, except I worry that at some point in the future someone will think the first one is redundant, remove it and introduce an off by one. At a minimum, please add a comment for this or optionally clean up the logic so it isn't needed. Even just this would be enough to replace the unconditional assignment at the end of the loop):
```
// we necessarily replaced the space with '\0' already when stop is true,
// adding a second one is easy and doesn't hurt anything.
output[output_idx] = '\0';
}
```

Reply:
This is not redundant unfortunately. Note that the loop is *not* broken out of. The break applies to the switch statement. We are terminating the string, so to speak, here and asking the loop to stop iterating but the `output_idx` counter is still advanced by the for loop header. We will only break on the *next* iteration.

The purpose of the NUL termination after the loop is to ensure we copy the trailing NUL byte that is *not* copied by the loop (since it is the terminating condition).

Reply:
I reworked the loop to remove the ambiguity.

Reply:
> This is not redundant unfortunately. Note that the loop is _not_ broken out of. The break applies to the switch statement. We are terminating the string, so to speak, here and asking the loop to stop iterating but the `output_idx` counter is still advanced by the for loop header. We will only break on the _next_ iteration.
> 
> The purpose of the NUL termination after the loop is to ensure we copy the trailing NUL byte that is _not_ copied by the loop (since it is the terminating condition).

I guess I wasn't clear in my comment. I was very much saying it was *not* redundant and important that it stay there, but that it might appear redundant to someone else 6 months from now, and I worried about the potential of it being removed.